In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LinearRegression, LogisticRegression, LassoCV, RidgeCV, Lasso, Ridge
from sklearn import metrics
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor 
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.ensemble import BaggingClassifier, BaggingRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor, ExtraTreesClassifier
from sklearn.ensemble import GradientBoostingClassifier, AdaBoostClassifier, AdaBoostRegressor, VotingClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV, train_test_split, StratifiedKFold

from sklearn.svm import LinearSVC, SVC, SVR

In [2]:
df = pd.read_csv('../data/numeric_text_combined.csv')

In [4]:
df_model = df

In [5]:
df_model.head()

Unnamed: 0  protestnumber  protesterviolence  participants stateresponse  \
0           0              1                0.0          3000        ignore   
1           1              2                0.0          1000        ignore   
2           2              3                0.0           500        ignore   
3           3              4                1.0           300  accomodation   
4           4              5                1.0           950       arrests   

   labor wage dispute  land farm issue  police brutality  \
0                   1                0                 0   
1                   0                0                 0   
2                   0                0                 0   
3                   0                1                 0   
4                   0                0                 0   

   political behavior, process  price increases, tax policy  ...  \
0                            1                            0  ...   
1                            1                            0  ...   
2                            1                            0  ...   
3                            0                            0  ...   
4                            1                            0  ...   

   yesterday morning  yesterday protest  yesterday thousand    young  \
0                0.0                0.0                 0.0  0.00000   
1                0.0                0.0                 0.0  0.36621   
2                0.0                0.0                 0.0  0.00000   
3                0.0                0.0                 0.0  0.00000   
4                0.0                0.0                 0.0  0.00000   

   young men  young people  youth  yugoslavia  zimbabwe  zone  
0        0.0       0.00000    0.0         0.0       0.0   0.0  
1        0.0       0.47086    0.0         0.0       0.0   0.0  
2        0.0       0.00000    0.0         0.0       0.0   0.0  
3        0.0       0.00000    0.0         0.0       0.0   0.0  
4        0.0       0.00000    0.0         0.0       0.0   0.0  

[5 rows x 3214 columns]

In [6]:
df_model.drop(columns = 'Unnamed: 0', inplace = True)

In [7]:
df_model.head()

protestnumber  protesterviolence  participants stateresponse  \
0              1                0.0          3000        ignore   
1              2                0.0          1000        ignore   
2              3                0.0           500        ignore   
3              4                1.0           300  accomodation   
4              5                1.0           950       arrests   

   labor wage dispute  land farm issue  police brutality  \
0                   1                0                 0   
1                   0                0                 0   
2                   0                0                 0   
3                   0                1                 0   
4                   0                0                 0   

   political behavior, process  price increases, tax policy  \
0                            1                            0   
1                            1                            0   
2                            1                            0   
3                            0                            0   
4                            1                            0   

   removal of politician  ...  yesterday morning  yesterday protest  \
0                      0  ...                0.0                0.0   
1                      0  ...                0.0                0.0   
2                      0  ...                0.0                0.0   
3                      0  ...                0.0                0.0   
4                      0  ...                0.0                0.0   

   yesterday thousand    young  young men  young people  youth  yugoslavia  \
0                 0.0  0.00000        0.0       0.00000    0.0         0.0   
1                 0.0  0.36621        0.0       0.47086    0.0         0.0   
2                 0.0  0.00000        0.0       0.00000    0.0         0.0   
3                 0.0  0.00000        0.0       0.00000    0.0         0.0   
4                 0.0  0.00000        0.0       0.00000    0.0         0.0   

   zimbabwe  zone  
0       0.0   0.0  
1       0.0   0.0  
2       0.0   0.0  
3       0.0   0.0  
4       0.0   0.0  

[5 rows x 3213 columns]

In [8]:
df_model.isnull().sum()

protestnumber         0
protesterviolence     0
participants          0
stateresponse         0
labor wage dispute    0
                     ..
young people          0
youth                 0
yugoslavia            0
zimbabwe              0
zone                  0
Length: 3213, dtype: int64

In [9]:
df_model.isnull().sum().sum()

0

No Nulls

In [10]:
df_model.shape

(14474, 3213)

### State Response Basic 

In [12]:
df_model['stateresponse'].value_counts(normalize = True)
# need to beat 53%

ignore             0.530883
crowd dispersal    0.174036
arrests            0.101285
accomodation       0.069918
killings           0.052992
beatings           0.042490
shootings          0.028396
Name: stateresponse, dtype: float64

In [13]:
X = df_model._get_numeric_data()

In [14]:
y = df_model['stateresponse']

In [15]:
X.columns

Index(['protestnumber', 'protesterviolence', 'participants',
       'labor wage dispute', 'land farm issue', 'police brutality',
       'political behavior, process', 'price increases, tax policy',
       'removal of politician', 'social restrictions',
       ...
       'yesterday morning', 'yesterday protest', 'yesterday thousand', 'young',
       'young men', 'young people', 'youth', 'yugoslavia', 'zimbabwe', 'zone'],
      dtype='object', length=3212)

#### Train Test Split Then StandardScaler

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [17]:
ss = StandardScaler()
X_train_sc = ss.fit_transform(X_train)
X_test_sc = ss.transform(X_test)

### Basic Models

In [18]:
abc = AdaBoostClassifier()
bc = BaggingClassifier()
rfc = RandomForestClassifier()
lr = LogisticRegression()
knc = KNeighborsClassifier()
svc = SVC()
dtc = DecisionTreeClassifier()
classification_models = [abc, bc, rfc, lr, knc, svc, dtc]
for model in classification_models:
    print(model)
    model.fit(X_train_sc, y_train)
    print(model.score(X_train_sc, y_train))
    print(model.score(X_test_sc, y_test))

AdaBoostClassifier()
0.6548134500230308
0.6413373860182371
BaggingClassifier()
0.9836020267157992
0.6637192594639403
RandomForestClassifier()
1.0
0.6736667587731417
LogisticRegression()


/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.9893136803316444
0.5355070461453441
KNeighborsClassifier()
0.46439428834638413
0.28847747996684164
SVC()
0.871672040534316
0.6402321083172147
DecisionTreeClassifier()
1.0
0.5844155844155844


### Comparing Pre_Clean vs Post_Clean
- AdaBoostClassifier()
      - Training: 0.5879318286503915 vs 0.6548134500230308
      - Test: 0.5816523901630285 vs 0.6413373860182371
- BaggingClassifier()
      - Training: 0.9831414094887149 vs 0.9836020267157992
      - Test: 0.5451782260292899 vs 0.6637192594639403
- RandomForestClassifier()
      - Training: 0.9997236296637494 vs 1.0
      - Test: 0.5852445426913512 vs 0.6736667587731417
- LogisticRegression()
      - Training: 0.5912482726853985 vs 0.9893136803316444
      - Test: 0.581928709588284  vs 0.5355070461453441
- KNeighborsClassifier()
      - Training: 0.6465223399355136 vs 0.46439428834638413
      - Test: 0.5078751036197845 vs 0.28847747996684164
- SVC()
      - Training: 0.5977890373099954 vs 0.871672040534316
      - Test: 0.5835866261398176 vs 0.6402321083172147
- DecisionTreeClassifier()
      - Training: 0.9998157531091663 vs 1.0
      - Test: 0.4340978170765405 vs 0.5844155844155844

## Remap State Response To Numeric Values, Check Scores
- Also will have to remap so gridsearching works

In [19]:
df_model['stateresponseweight'] = df_model['stateresponse'].map({'accomodation': 2, 'ignore': 1, 'crowd dispersal': 3,
                                                       'arrests': 4, 'beatings': 5, 'shootings': 6, 'killings': 7})

In [24]:
X = df_model._get_numeric_data().drop(columns = 'stateresponseweight')
y = df_model['stateresponseweight']

In [25]:
X.columns

Index(['protestnumber', 'protesterviolence', 'participants',
       'labor wage dispute', 'land farm issue', 'police brutality',
       'political behavior, process', 'price increases, tax policy',
       'removal of politician', 'social restrictions',
       ...
       'yesterday morning', 'yesterday protest', 'yesterday thousand', 'young',
       'young men', 'young people', 'youth', 'yugoslavia', 'zimbabwe', 'zone'],
      dtype='object', length=3212)

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [27]:
ss = StandardScaler()
X_train_sc = ss.fit_transform(X_train)
X_test_sc = ss.transform(X_test)

In [29]:
abc = AdaBoostClassifier()
bc = BaggingClassifier()
rfc = RandomForestClassifier()
lr = LogisticRegression()
knc = KNeighborsClassifier()
svc = SVC()
dtc = DecisionTreeClassifier()
classification_models = [abc, bc, rfc, lr, knc, svc, dtc]
for model in classification_models:
    print(model)
    model.fit(X_train_sc, y_train)
    print(model.score(X_train_sc, y_train))
    print(model.score(X_test_sc, y_test))

AdaBoostClassifier()
0.6536158452326117
0.649626968775905
BaggingClassifier()
0.9736526946107784
0.6681403702680299
RandomForestClassifier()
1.0
0.6769825918762089
LogisticRegression()


/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.9891294334408107
0.547112462006079
KNeighborsClassifier()
0.5995393827729157
0.46200607902735563
SVC()
0.8699216950713957
0.642718983144515
DecisionTreeClassifier()
1.0
0.5822050290135397


### Comparing Weighted vs Post_Clean
- AdaBoostClassifier()
      - Training: 0.6536158452326117 vs 0.6548134500230308
      - Test: 0.649626968775905 vs 0.6413373860182371
- BaggingClassifier()
      - Training: 0.9736526946107784 vs 0.9836020267157992
      - Test: 0.6747720364741642 vs 0.6681403702680299
- RandomForestClassifier()
      - Training: 1.0 vs 1.0
      - Test: 0.6769825918762089 vs 0.6736667587731417
- LogisticRegression()
      - Training: 0.9891294334408107 vs 0.9893136803316444
      - Test: 0.547112462006079  vs 0.5355070461453441
- KNeighborsClassifier()
      - Training: 0.5995393827729157 vs 0.46439428834638413
      - Test: 0.46200607902735563 vs 0.28847747996684164
- SVC()
      - Training: 0.8699216950713957 vs 0.871672040534316
      - Test: 0.642718983144515 vs 0.6402321083172147
- DecisionTreeClassifier()
      - Training: 1.0 vs 1.0
      - Test: 0.5822050290135397 vs 0.5844155844155844

## GridSearch / HyperParameters

### Random Forest 

In [31]:
rfc = RandomForestClassifier()
rf_params = {
    'n_estimators': [100, 150, 200],
    'min_samples_leaf': [1, 5, 10],
    'max_depth': [None, 1, 3, 5],
}
gs = GridSearchCV(rfc, param_grid=rf_params, cv=5)
gs.fit(X_train_sc, y_train)
print(gs.best_score_)
gs.best_params_

0.6687240902809766


{'max_depth': None, 'min_samples_leaf': 1, 'n_estimators': 200}

In [32]:
gs.score(X_train_sc, y_train), gs.score(X_test_sc, y_test)

(0.9555043758636573, 0.6758773141751865)

### Want to Try out Extra Tree

In [36]:
et = ExtraTreesClassifier()
et.fit(X_train_sc, y_train)
cross_val_score(et, X_train, y_train, cv=5).mean()

0.6654076462459696

In [37]:
et.score(X_train_sc, y_train), et.score(X_test_sc, y_test)

(1.0, 0.6717325227963525)

### SVM

In [41]:
svc = SVC()
pgrid = {'kernel': ['rbf', 'sigmoid']}
cv = StratifiedKFold(n_splits=3, shuffle=True)

gcv = GridSearchCV(svc, pgrid, cv=cv)
gcv.fit(X_train_sc, y_train)
gcv.best_params_
gcv.best_score_

0.6479043638459325

In [42]:
gcv.score(X_train_sc, y_train), gcv.score(X_test_sc, y_test)

(0.7174573929064947, 0.6634429400386848)

### Regularization

In [38]:
ridge_model = Ridge(alpha=10)
ridge_model.fit(X_train_sc, y_train)
print(ridge_model.score(X_train_sc, y_train))
print(ridge_model.score(X_test_sc, y_test))

0.6842535106917533
0.3710010993078108


In [39]:
lasso = Lasso()
lasso.fit(X_train_sc, y_train)
print(lasso.score(X_train_sc, y_train))
print(lasso.score(X_test_sc, y_test))

0.0
-7.251163992139098e-06
